src/package-hoister.js

```diff
@@ -242,7 +242,17 @@ export default class PackageHoister {
           continue;
         }
 
-        const isMarkedAsOptional = depinfo.pkg._reference && depinfo.pkg._reference.optional && this.ignoreOptional;
+        const depRef = depinfo.pkg._reference;
+
+        let isMarkedAsOptional = depRef && depRef.optional && this.ignoreOptional;
+
+        // If it's marked as optional, but the parent is required and the
+        // dependency was not listed in `optionalDependencies`, then we mark the
+        // dependency as required.
+        if (isMarkedAsOptional && info.isRequired && depRef && depRef.hint !== 'optional') {
```

Comment:
Can we not use `depRef.isRequired` instead of this? Or may be add `depRef.isOptional`?

Reply:
I tried that at first, but it fails with "deeper optionals". E.g.

```
foo (in `optionalDependencies`) -> bar -> baz
```

Here both `bar` and `baz` will have `_reference.isOptional === true`. However, as this code doesn't mark the reference when performing this logic (only setting `depinfo.isRequired = true`), `baz` will stay `optional` as `bar._reference.optional === true`. If we changed the logic to also do a `depRef.setOptional(false)` in addition to setting `depinfo.isRequired` this approach would work. That was my initial fix, but I wasn't sure about doing that work from the package hoister. Thoughts?